File: backend/src/main/java/com/example/mypixel/model/node/MedianBlurNode.java
package com.example.mypixel.model.node;

import com.example.mypixel.exception.InvalidNodeParameter;
import com.example.mypixel.model.Parameter;
import com.example.mypixel.model.ParameterType;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import org.springframework.lang.NonNull;

import java.util.HashSet;
import java.util.Map;

@MyPixelNode("MedianBlur")
public class MedianBlurNode extends Node {

    @JsonCreator
    public MedianBlurNode(
            @JsonProperty("id") @NonNull Long id,
            @JsonProperty("type") @NonNull String type,
            @JsonProperty("inputs") Map<String, Object> inputs) {
        super(id, type, inputs);
    }

    @Override
    public Map<String, Parameter> getInputTypes() {
        return Map.of(
                "files", Parameter.required(ParameterType.FILEPATH_ARRAY),
                "ksize", Parameter.required(ParameterType.INT)
        );
    }

    @Override
    public Map<String, Object> getDefaultInputs() {
        return Map.of(
                "files", new HashSet<String>(),
                "ksize", 3
        );
    }

    @Override
    public Map<String, Parameter> getOutputTypes() {
        return Map.of("files", Parameter.required(ParameterType.FILEPATH_ARRAY));
    }

    @Override
    public Map<String, String> getDisplayInfo() {
        return Map.of(
                "category", "Filtering",
                "description", "Blurs an image using the specified kernel size",
                "color", "#FF8A65",
                "icon", "BlurIcon"
        );
    }

    @Override
    public Map<String, Object> exec() {
        HashSet<String> files = (HashSet<String>) inputs.get("files");
        Map<String, Object> outputs;

        int ksize = (int) inputs.get("ksize");

        batchProcessor.processBatches(files,
                filepath -> filteringService.medianBlur(filepath, ksize));

        outputs = Map.of("files", files);

        return outputs;
    }

    @Override
    public void validate() {
        int ksize = (int) inputs.get("ksize");

        if (ksize < 2 || ksize % 2 == 0) {
            throw new InvalidNodeParameter("KSize must be greater than 1 and odd");
        }
    }
}

File: node/nodes/median_blur_node.py
from typing import Dict, Any

from metadata import Metadata
from node import Node
from storage_client import StorageClient


class MedianBlurNode(Node):

    node_type = "MedianBlur"

    def get_input_types(self) -> Dict[str, Dict[str, Any]]:
        return {
            "files": {
                "type": "FILEPATH_ARRAY",
                "required": True,
                "widget": "LABEL",
                "default": set()
            },
            "ksize": {
                "type": "INT",
                "required": True,
                "widget": "INPUT",
                "default": 3
            }
        }

    def get_output_types(self) -> Dict[str, Dict[str, Any]]:
        return {
            "files": {
                "type": "FILEPATH_ARRAY",
                "required": True,
                "widget": "LABEL"
            }
        }

    def get_display_info(self) -> Dict[str, str]:
        return {
            "category": "Filtering",
            "description": "Blurs an image using the specified kernel size",
            "color": "#FF8A65",
            "icon": "BlurIcon"
        }

    def exec(self, files, ksize, meta: Metadata) -> Dict[str, Any]:
        output_files = []

        for file in files:
            output_files.append(StorageClient.store_from_workspace_to_task(meta.task_id, meta.id, file))


        return {"files": output_files}

    def validate(self, files, ksize, meta) -> None:
        try:
            ksize = int(ksize)
        except (TypeError, ValueError):
            raise ValueError("ksize must be an integer")

        if ksize < 2 or ksize % 2 == 0:
            raise ValueError("KSize must be greater than 1 and odd")
